### PR TITLE
refactor: extract proto map processing to Config::Utility and use absl::flat_hash_map for storage.

### PR DESCRIPTION
**Pull request by lizan**
Signed-off-by: Lizan Zhou <lizan@tetrate.io>

*Description*:
For #4475. Not very clean, but will make add google.protobuf.Any support easier.

*Risk Level*: Low
*Testing*: CI (refactor shouldn't affect existing tests)
*Docs Changes*: N/A
*Release Notes*: N/A

### CODE REVIEW

File: source/common/config/utility.h [hunk 1]
```diff
@@ -275,45 +277,32 @@ class Utility {
...
-   */
-  static ProtobufTypes::MessagePtr
-  translateToFactoryProtocolOptionsConfig(const Protobuf::Message& source, const std::string& name,
-                                          Server::Configuration::ProtocolOptionsFactory& factory) {
-    ProtobufTypes::MessagePtr config = factory.createEmptyProtocolOptionsProto();
-
-    if (config == nullptr) {
-      throw EnvoyException(fmt::format("filter {} does not support protocol options", name));
+  template <class ConfigObject, class Factory>
+  static absl::flat_hash_map<std::string, ConfigObject> translateOpaqueConfigMap(
```
Comment (htuch) [inline source/common/config/utility.h]: I think this callback oriented approach is kind of hard to follow. Could we have a virtual interface at the expense of it being a bit more verbose? Otherwise, this starts to look like Javascript :)

Comment (lizan) [reply source/common/config/utility.h]: Yeah I know this is not very clean, but a virtual interface will be templated and introduce much more boiler plate so it might not worth. Another option is do those as static functions for `ConfigObject` and use template. Thought it is hard to follow anyway. WDYT?

I'm fine to revert the refactor part and just introduce flat_hash_map too.

Comment (jmarantz) [reply source/common/config/utility.h]: Introducing flat_hash_map separately seems like a win.

BTW I know the general recommendation is to use flat_hash_map for string, but I didn't find that to be optimal. It might be good to try node_hash_map and see which is faster/smaller.

You can't really go too wrong with flat_hash_map, but the usual motivation to switch is to improve performance it might be good to measure it.

OTOH if this isn't performance sensitive at all and you are just doing it to be consistent then carry on :)

Comment (lizan) [reply source/common/config/utility.h]: @jmarantz Per https://abseil.io/blog/20180927-swisstables

> We generally recommend that you use absl::flat_hash_map<K, std::unique_ptr<V>> instead of absl::node_hash_map<K, V>.

So I think we could go with flat_hash_map as the ConfigObject is wrapped in const shared ptr.

The write part (this part) isn't performance critical, the getter part is on the data path.

Comment (jmarantz) [reply source/common/config/utility.h]: Yup should work fine with that value. I just found it was slower/bigger than node_hash_map due to the complexity of the std;:string as key. Works fine though, and if perf is not a consideration then this is fine.

Comment (htuch) [reply source/common/config/utility.h]: @lizan I think just doing some virtual class hierarchy with the concrete types and then using them in the template concretely with the instantiated types? I agree that in general, virtual inheritance and templates don't work well.

I'm in the Anything But Callbacks Everywhere camp :)

**Comment by htuch** [comment]
@lizan friendly ping :)

**Comment by lizan** [comment]
Closing this since it seems to be too much effort for just two use cases, adding Any to each of them is not that bad. Will do flat_hash_map in another PR.
